src/Analyzers/CSharp/Analyzers/NamespaceSync/CSharpNamespaceSyncDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Linq;
+using Microsoft.CodeAnalysis.Analyzers.NamespaceSync;
+using Microsoft.CodeAnalysis.CSharp.LanguageServices;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+
+namespace Microsoft.CodeAnalysis.CSharp.Analyzers.NamespaceSync
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class CSharpNamespaceSyncDiagnosticAnalyzer : AbstractNamespaceSyncDiagnosticAnalyzer<NamespaceDeclarationSyntax>
+    {
+        private static readonly LocalizableResourceString s_localizableTitle = new LocalizableResourceString(
+          nameof(CSharpAnalyzersResources.Namespace_does_not_match_folder_structure), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources));
+
+        private static readonly LocalizableResourceString s_localizableInsideMessage = new LocalizableResourceString(
+            nameof(CSharpAnalyzersResources.Namespace_0_does_not_match_folder_structure_expected_1), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources));
+
+        public CSharpNamespaceSyncDiagnosticAnalyzer()
+            : base(s_localizableTitle,
+                s_localizableInsideMessage)
+        {
+        }
+
+        protected override ISyntaxFacts GetSyntaxFacts() => CSharpSyntaxFacts.Instance;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeNamespaceNode, SyntaxKind.NamespaceDeclaration);
+        }
+
+        protected override bool ContainsPartialTypeWithMultipleDeclarations(NamespaceDeclarationSyntax namespaceDeclaration, SemanticModel semanticModel)
+        {
+            var partialMemberDecls = namespaceDeclaration
+                .Members.OfType<TypeDeclarationSyntax>()
+                .Where(t => t.Modifiers.Any(SyntaxKind.PartialKeyword));
+
+            foreach (var memberDecl in partialMemberDecls)
+            {
+                var memberSymbol = semanticModel.GetDeclaredSymbol(memberDecl);
+
+                // Simplify the check by assuming no multiple partial declarations in one document
```

Comment:
Does `DeclaringSyntaxReferences` includes generated code? i.e. if user declared a partial type in source and expect code generator to generate the the other part of the partial type.

Reply:
That's a good question. I wonder if we can construct a unit test for that... Could easily make a source generator that just drops a known partial type